distribution/xfer/download.go

```diff
@@ -70,7 +86,7 @@ type DownloadDescriptor interface {
 	// before).
 	DiffID() (layer.DiffID, error)
 	// Download is called to perform the download.
-	Download(ctx context.Context, progressOutput progress.Output) (io.ReadCloser, int64, error)
+	Download(ctx context.Context, progressOutput progress.Output, cacheDir string) (io.ReadCloser, int64, error)
```

Comment:
@stevvooe unfortunately it's needed. 

cacheDir is instantiated here: https://github.com/docker/docker/pull/28348/files#diff-0c4769e1568c0f439863abd30df1ceccR264 from transferKey

And transferKey is the concatenation of the layer descriptor's keys: https://github.com/docker/docker/pull/28348/files#diff-0c4769e1568c0f439863abd30df1ceccR181

So the descriptor alone is not enough.